File: Codes/Programmers/거리두기확인하기.py
from collections import deque

DELTA = [(1, 0), (0, 1), (0, -1), (-1, 0)]


def is_anyone_nearby(place, visited, sy, sx):
    queue = deque([(sy, sx)])
    for _ in range(2):
        temp = deque()
        while queue:
            y, x = queue.popleft()
            visited[y][x] = True
            for dy, dx in DELTA:
                ny, nx = y + dy, x + dx
                if ny < 0 or ny >= 5 or nx < 0 or nx >= 5:
                    continue
                if visited[ny][nx] or place[ny][nx] == "X":
                    continue
                temp.append((ny, nx))
                if place[ny][nx] == "P":
                    return True

        queue = temp


def check(place):
    visited = [[False] * 5 for _ in range(5)]
    for i in range(5):
        for j in range(5):
            if place[i][j] == "P":
                if is_anyone_nearby(place, visited, i, j):
                    return 0
    return 1


def solution(places):
    return [check(place) for place in places]

File: Codes/Programmers/거리두기확인하기.js
function check(place) {
  const visited = Array.from(Array(5), () => Array(5).fill(false));
  const delta = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
  ];

  function visit_two(sy, sx) {
    let queue = [[sy, sx]];
    visited[sy][sx] = true;

    for (let i = 0; i < 2; i++) {
      const temp = [];
      while (queue.length > 0) {
        const [y, x] = queue.shift();
        for (let d of delta) {
          const ny = y + d[0];
          const nx = x + d[1];

          if (ny < 0 || ny >= 5 || nx < 0 || nx >= 5 || visited[ny][nx]) {
            continue;
          }
          if (place[ny][nx] === "X") {
            continue;
          }
          if (place[ny][nx] === "P") {
            return true;
          }
          visited[ny][nx] = true;
          temp.push([ny, nx]);
        }
      }
      queue = temp;
    }
    return false;
  }

  for (let i = 0; i < 5; i++) {
    for (let j = 0; j < 5; j++) {
      if (place[i][j] === "P" && !visited[i][j]) {
        if (visit_two(i, j)) {
          return 0;
        }
      }
    }
  }
  return 1;
}

function solution(places) {
  return places.map((place) => check(place));
}
